### PR TITLE
Feature/remove whisper common hallucinations

### PR DESCRIPTION
**Pull request by kkzxak47**
image
image

## Summary by Sourcery

Implements a hallucination cleaner for Whisper transcriptions to remove common phrases that the model often hallucinates at the end of transcriptions. It uses spaCy's vector similarity to detect and remove these phrases.

New Features:
- Introduces a hallucination cleaner for Whisper transcriptions using spaCy's vector similarity to remove common phrases.

Tests:
- Adds tests for the WhisperHallucinationCleaner class, including tests for initialization, NLP loading, similarity checking, sentence splitting, text cleaning, and edge cases.

## Summary by Sourcery

Implements a hallucination cleaner for Whisper transcriptions to remove common phrases that the model often hallucinates at the end of transcriptions by using spaCy's vector similarity to detect and remove these phrases. It also adds a setting to enable or disable the hallucination cleaner.

New Features:
- Introduces a hallucination cleaner for Whisper transcriptions using spaCy's vector similarity to remove common phrases.
- Adds a setting to enable or disable the hallucination cleaner in the settings window.

Tests:
- Adds tests for the WhisperHallucinationCleaner class, including tests for initialization, NLP loading, similarity checking, sentence splitting, text cleaning, and edge cases.

### CODE REVIEW

File: src/FreeScribe.client/services/whisper_hallucination_cleaner.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        # First check for exact matches (case insensitive)
+        if any(h in sentence for h in self.hallucinations):
+            logger.debug(f"Sentence contains a hallucination: {sentence}")
+            return True
+            
+        # Process the sentence
+        sentence_doc = self.nlp(sentence)
+        
+        # Longer sentences are less likely to be hallucinations
+        if len(sentence_doc) > MAX_SENTENCE_LENGTH:
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply src/FreeScribe.client/services/whisper_hallucination_cleaner.py:202]: fixed

File: src/FreeScribe.client/client.py [hunk 2]
```diff
@@ -1796,8 +1799,13 @@ def faster_whisper_transcribe(audio):
...
-        return "".join(f"{segment.text} " for segment in segments)
+        result = "".join(f"{segment.text} " for segment in segments)
+        logger.debug(f"Result: {result}")
+        cleaned_result = hallucination_cleaner.clean_text(result)
+        logger.debug(f"Cleaned result: {cleaned_result}")
+        return cleaned_result
     except Exception as e:
+        logger.exception(f"Error during transcription: {str(e)}")
         error_message = f"Transcription failed: {str(e)}"
         print(f"Error during transcription: {str(e)}")
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply src/FreeScribe.client/client.py:1810]: removed

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+    assert "This is a normal sentence" in cleaned
+    assert "This is another normal sentence" in cleaned
+    
+    # Test with empty text
+    assert cleaner.clean_text("") == ""
+    
+    # Test with text containing no hallucinations
+    text = "This is a normal sentence. This is another normal sentence."
+    cleaned = cleaner.clean_text(text)
+    assert cleaned.replace("..", ".") == text  # Account for potential double periods
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:95]: fixed

File: src/FreeScribe.client/services/whisper_hallucination_cleaner.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
+"""Whisper Hallucination Cleaner.
+
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply src/FreeScribe.client/services/whisper_hallucination_cleaner.py:2]: done

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+    similar_phrases = [
+        "thanks for your time",
+        "appreciate your watching",
+        "see you in our next video",
+        "goodbye everyone"
+    ]
+    
+    for phrase in similar_phrases:
+        # These should be caught by vector similarity
+        assert cleaner._is_similar_to_hallucination(phrase)
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:228]: I think this one is ok, it is simple and strictly for testing purposes, it's just looping through cases and run test.

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+    
+    monkeypatch.setattr(spacy.util, "is_package", mock_is_package)
+    monkeypatch.setattr(spacy.cli, "download", mock_download)
+    
+    if success:
+        assert download_spacy_model()
+        assert len(calls) == 1
+    else:
+        assert not download_spacy_model()
+        assert len(calls) == 3  # Should try 3 times
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:275]: Similarly, It is simple and pretty straight forward.

File: src/FreeScribe.client/services/whisper_hallucination_cleaner.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+    def __init__(self, similarity_threshold: float = SIMILARITY_THRESHOLD):
+        """Initialize the cleaner with a similarity threshold.
+        
+        :param similarity_threshold: The minimum similarity ratio (0-1) between a sentence
+                                   and a hallucination to consider it a match
+        :type similarity_threshold: float
+        """
+        self.similarity_threshold = similarity_threshold
+        # Store hallucinations as a set for O(1) membership tests
+        self.hallucinations = set(h.lower() for h in COMMON_HALUCINATIONS)
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply src/FreeScribe.client/services/whisper_hallucination_cleaner.py:130]: done

File: src/FreeScribe.client/services/whisper_hallucination_cleaner.py [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+        
+        :param text: Text to normalize
+        :type text: str
+        :returns: Normalized text
+        :rtype: str
+        """
+        # Remove punctuation and normalize whitespace
+        text = text.strip().lower()
+        text = ' '.join(text.split())  # Normalize whitespace
+        return text
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply src/FreeScribe.client/services/whisper_hallucination_cleaner.py:174]: done

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+def test_initialization(cleaner):
+    """Test the initialization of WhisperHallucinationCleaner.
+    
+    :param cleaner: The WhisperHallucinationCleaner fixture
+    :type cleaner: WhisperHallucinationCleaner
+    """
+    assert cleaner.similarity_threshold == SIMILARITY_THRESHOLD
+    assert isinstance(cleaner.hallucinations, set)
+    # Both should already be sets of lowercase strings
+    assert cleaner.hallucinations == set(h.lower() for h in COMMON_HALUCINATIONS)
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:24]: done

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+        * Proper handling of case-insensitive duplicates
+        * Caching behavior
+    
+    :param cleaner: The WhisperHallucinationCleaner fixture
+    :type cleaner: WhisperHallucinationCleaner
+    """
+    # Test that docs are created correctly
+    docs = cleaner.hallucination_docs
+    # Get unique lowercase hallucinations since some might be duplicates when lowercased
+    unique_hallucinations = len(set(h.lower() for h in COMMON_HALUCINATIONS))
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:179]: done

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+    
+    # Test phrases that shouldn't be similar
+    dissimilar_phrases = [
+        "the weather is nice today",
+        "please pass the salt",
+        "what time is the meeting"
+    ]
+    
+    for phrase in dissimilar_phrases:
+        assert not cleaner._is_similar_to_hallucination(phrase)
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:243]: It's ok. wont fix

File: src/FreeScribe.client/services/whisper_hallucination_cleaner.py [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+    def _is_similar_to_hallucination(self, sentence: str) -> bool:
+        """Check if a sentence is similar to any known hallucination using vector similarity.
+        
+        :param sentence: The sentence to check
+        :type sentence: str
+        :returns: True if the sentence is similar to a hallucination, False otherwise
+        :rtype: bool
+        """
+        if not sentence:
+            return False
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply src/FreeScribe.client/services/whisper_hallucination_cleaner.py:211]: True, we should keep gathering feedback. Common hallucinations are mostly short sentences that carry little meaning for note generation.

File: client_requirements_nvidia.txt [hunk 13]
```diff
@@ -71,3 +72,4 @@ nvidia-cuda-nvrtc-cu12==12.4.127
 nvidia-cublas-cu12==12.4.5.8
 psutil==6.1.0
 Pillow==10.2.0
+en_core_web_md @ https://github.com/explosion/spacy-models/releases/download/en_core_web_md-3.7.1/en_core_web_md-3.7.1-py3-none-any.whl
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply client_requirements_nvidia.txt:75]: No, it is not a mismatch.
image

File: src/FreeScribe.client/services/whisper_hallucination_cleaner.py [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+    "bye",
+]
+
+# Calculate max length based on tokens instead of characters for more accurate comparison
+MAX_SENTENCE_LENGTH = max(len(hallucination.split()) for hallucination in COMMON_HALUCINATIONS)
+SIMILARITY_THRESHOLD = 0.9
+SPACY_MODEL_NAME = "en_core_web_md"
+
+
+def download_spacy_model():
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply src/FreeScribe.client/services/whisper_hallucination_cleaner.py:77]: 1. You are contradicting yourself: ```The download_spacy_model function could be improved by using subprocess.run instead of spacy.cli.download to better handle errors and output.```
2. Good point. Done.

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+    similar_phrases = [
+        "thanks for your time",
+        "appreciate your watching",
+        "see you in our next video",
+        "goodbye everyone"
+    ]
+    
+    for phrase in similar_phrases:
+        # These should be caught by vector similarity
+        assert cleaner._is_similar_to_hallucination(phrase)
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:233]: This one is justified. It's natural to loop through cases and test.

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+    
+    # Test phrases that shouldn't be similar
+    dissimilar_phrases = [
+        "the weather is nice today",
+        "please pass the salt",
+        "what time is the meeting"
+    ]
+    
+    for phrase in dissimilar_phrases:
+        assert not cleaner._is_similar_to_hallucination(phrase)
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:243]: likewise

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+        # Create a mock CompletedProcess object
+        class MockCompletedProcess:
+            def __init__(self, returncode, stderr=""):
+                self.returncode = returncode
+                self.stderr = stderr
+                
+        if success:
+            return MockCompletedProcess(0)
+        else:
+            return MockCompletedProcess(1, "Download failed")
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:278]: This is a trivial case.

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+    
+    monkeypatch.setattr(spacy.util, "is_package", mock_is_package)
+    monkeypatch.setattr("subprocess.run", mock_run)
+    
+    if success:
+        assert download_spacy_model()
+        assert len(calls) == 1  # Should be called once on success
+    else:
+        assert not download_spacy_model()
+        assert len(calls) == 3  # Should try 3 times on failure
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:288]: This is als trivia

File: client_requirements_nvidia.txt (hunk 13 above)
[the remark this answers is not in the document]
Comment (kkzxak47) [reply client_requirements_nvidia.txt:75]: You are hallucinating.

[the remark this answers is not in the document]
Comment (kkzxak47) [reply client_requirements_nvidia.txt:75]: SpaCy models cannot be installed using syntax like "en_core_web_md==3.7.1", this is the explicit way to pin the version.

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+    similar_phrases = [
+        "thanks for your attention",
+        "thank you for listening",
+        "see you in another one",
+        "Thanks for watching, and I'll see you in the next video, and I'll see you in the next video.",
+    ]
+    
+    for phrase in similar_phrases:
+        # These should be caught by vector similarity
+        assert cleaner._is_similar_to_hallucination(phrase)
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:233]: ok, converting loops to fixture

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 20]
```diff
@@ -0,0 +1,300 @@
...
+    
+    # Test phrases that shouldn't be similar
+    dissimilar_phrases = [
+        "the weather is nice today",
+        "please pass the salt",
+        "what time is the meeting"
+    ]
+    
+    for phrase in dissimilar_phrases:
+        assert not cleaner._is_similar_to_hallucination(phrase)
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:243]: done

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 21]
```diff
@@ -0,0 +1,300 @@
...
+        # Create a mock CompletedProcess object
+        class MockCompletedProcess:
+            def __init__(self, returncode, stderr=""):
+                self.returncode = returncode
+                self.stderr = stderr
+                
+        if success:
+            return MockCompletedProcess(0)
+        else:
+            return MockCompletedProcess(1, "Download failed")
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:278]: ok, converting if to fixture

[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:278]: done

File: tests/services/test_whisper_hallucination_cleaner.py [hunk 22]
```diff
@@ -0,0 +1,300 @@
...
+    
+    monkeypatch.setattr(spacy.util, "is_package", mock_is_package)
+    monkeypatch.setattr("subprocess.run", mock_run)
+    
+    if success:
+        assert download_spacy_model()
+        assert len(calls) == 1  # Should be called once on success
+    else:
+        assert not download_spacy_model()
+        assert len(calls) == 3  # Should try 3 times on failure
```
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py:288]: done

File: tests/services/test_whisper_hallucination_cleaner.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py]: loops and conditionals are removed

File: tests/services/test_whisper_hallucination_cleaner.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py]: fixed

File: tests/services/test_whisper_hallucination_cleaner.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py]: fixed

File: tests/services/test_whisper_hallucination_cleaner.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py]: fixed

File: tests/services/test_whisper_hallucination_cleaner.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (kkzxak47) [reply tests/services/test_whisper_hallucination_cleaner.py]: fixed

**Review by yingbull — APPROVED** [review]

**Review by yingbull — APPROVED** [review]
